torch/_dynamo/variables/builtin.py

```diff
@@ -1173,6 +1174,45 @@ def call_setattr(
             tx.output.side_effects.is_attribute_mutation(obj)
             and name_var.is_python_constant()
         ):
+            if isinstance(obj, variables.TensorVariable):
+                from .builder import wrap_fx_proxy
+
+                if name_var.as_python_constant() == "data":
+                    # Remove the old reference in tracked fakes - if we don't do this
+                    # new .data value size and shape differences will cause
+                    # tracked fakes to produce incorrect guards. This is sound because the TensorVariable
+                    # coming out of set_() below will be a new one, and get
+                    # installed in tracked fakes.
+                    to_remove = []
+                    for tf in tx.output.tracked_fakes:
+                        if tf.source == obj.source:
+                            to_remove.append(tf)
+                    for tf in to_remove:
+                        tx.output.tracked_fakes.remove(tf)
+
+                    # Step 1 - disable grad
+                    version = obj.as_proxy().node.meta["example_value"]._version
+                    with dynamo_disable_grad(tx), torch.no_grad():
+                        # Step 2 - call `set_`
+                        out = wrap_fx_proxy(
+                            tx,
+                            tx.output.create_proxy(
+                                "call_function",
+                                torch.Tensor.set_,
+                                *proxy_args_kwargs([obj, val], {}),
+                            ),
+                        )
+                    # Step 3 - drop the version counter - this is a hack required to get
+                    # .data setting to play correctly with the autograd engine.
+                    tx.output.create_proxy(
+                        "call_function",
+                        torch._C._autograd._unsafe_set_version_counter,
+                        (out.as_proxy(), 0),
+                        {},
+                    )
+                    # Step 4 - replace all reference to the current object with the new one
+                    tx.replace_all(obj, out)
```

Comment:
Although `set_` is supposed to be in-place, could we just add a test that the tensor object aliasing is preserved?

`python
z = x
x.data = y
x.add_(1)
return x is z
```

Reply:
That doesn't work with aot_autograd, but it does work with eager. Talking to @bdhirsh now. See the stacked PR.

I can change this to be its own test for eager only, and that should work.

Reply:
Right, I updated to just do `y.data = torch.zeros([0])`

Reply:
Can add :)